services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/multipart/MultipartDownloaderSubscriber.java

```diff
@@ -27,86 +27,115 @@
         this.s3 = s3;
         this.getObjectRequest = getObjectRequest;
-        this.responseFuture = new CompletableFuture<>();
     }
 
     @Override
     public void onSubscribe(Subscription s) {
-        if (this.subscription == null) {
-            this.subscription = s;
+        if (this.subscription != null) {
+            s.cancel();
+            return;
         }
-        s.request(1);
+        this.subscription = s;
+        this.subscription.request(1);
     }
 
     @Override
     public void onNext(AsyncResponseTransformer<GetObjectResponse, GetObjectResponse> asyncResponseTransformer) {
-        int part = currentPart.incrementAndGet();
-        if (totalPartKnown.get()) {
-            log.info(() -> String.format("total part: %s, current part: %s", totalParts.get(), part));
-        } else {
-            log.info(() -> String.format("part %s", part));
+        if (asyncResponseTransformer == null) {
+            throw new NullPointerException("onNext must not be called with null asyncResponseTransformer");
         }
-        if (totalPartKnown.get() && part > totalParts.get()) {
-            log.info(() -> "no more parts available, stopping");
+
+        int nextPartToGet = completedParts.get() + 1;
+        if (totalPartKnown.get() && nextPartToGet > totalParts.get()) {
             subscription.cancel();
             return;
         }
-        GetObjectRequest actualRequest = this.getObjectRequest.copy(req -> req.partNumber(part));
+
+        GetObjectRequest actualRequest = getObjectRequest.copy(req -> req.partNumber(nextPartToGet));
         CompletableFuture<GetObjectResponse> future = s3.getObject(actualRequest, asyncResponseTransformer);
-        future.whenComplete((response, e) -> {
-            if (e != null) {
-                responseFuture.completeExceptionally(e);
+        future.whenComplete((response, error) -> {
+            if (error != null) {
+                onError(error);
                 return;
             }
-            completed.incrementAndGet();
-            returnResponse = response;
-            log.info(() -> String.format("received '%s'", response.contentRange()));
+            int totalComplete = completedParts.incrementAndGet();
+            log.info(() -> String.format("completed part: %s", totalComplete));
+
             Integer partCount = response.partsCount();
-            if (totalPartKnown.compareAndSet(false, true)) {
+            if (partCount != null && totalPartKnown.compareAndSet(false, true)) {
+                log.info(() -> String.format("total parts: %s", partCount));
                 totalParts.set(partCount);
                 totalPartKnown.set(true);
             }
-            log.info(() -> String.format("total parts: %s", partCount));
-            if (totalParts.get() > 1) {
+            if (partCount != null && requiresMultipart()) {
                 subscription.request(1);
+            } else {
+                subscription.cancel();
             }
         });
```

Comment:
This is required, as the `MultipartDownloaderSubscriber` is the class tracking the total amount of parts and the current part. we need to call `cancel()` on the subscription for completing the upstream  `AsyncRresponseTransformer` (the one that was split) once everything is done.

Reply:
Oh I think I misunderstood. So we're essentially subscribing to an infinite publisher of `AsyncResponseTransformer`? I was thinking we could just let the `Publisher` dictate the total number of calls needed but if that's not the case then yeah you can disregard this.

Reply:
We do need to cancel the subscription once we reach the last part. The DownloaderSubscriber is the class tracking the total number of parts so that `SplittingTransformer` is not coupled with any service specific logic.

Reply:
Yeah, I think I prefer having the Publisher not coupled with any service specific logic, so it is this subscriber that would be tracking the amount of parts and requesting `AsyncResponseTransformer` from the publisher until the last part is reached